src/ir/memory-utils.cpp

```diff
@@ -37,10 +38,33 @@ bool flatten(Module& wasm) {
       return std::make_unique<Scanner>(noticesSegmentIdentity);
     }
 
-    void visitMemoryInit(MemoryInit* curr) { noticesSegmentIdentity = true; }
-    void visitDataDrop(DataDrop* curr) { noticesSegmentIdentity = true; }
-    void visitArrayNewData(ArrayNewData* curr) {
-      noticesSegmentIdentity = true;
+    void visitExpression(Expression* curr) {
+#define DELEGATE_ID curr->_id
+
+#define DELEGATE_START(id) [[maybe_unused]] auto* cast = curr->cast<id>();
+
+#define DELEGATE_GET_FIELD(id, field) cast->field
+
+#define DELEGATE_FIELD_TYPE(id, field)
+#define DELEGATE_FIELD_HEAPTYPE(id, field)
+#define DELEGATE_FIELD_CHILD(id, field)
+#define DELEGATE_FIELD_OPTIONAL_CHILD(id, field)
+#define DELEGATE_FIELD_INT(id, field)
+#define DELEGATE_FIELD_INT_ARRAY(id, field)
+#define DELEGATE_FIELD_LITERAL(id, field)
+#define DELEGATE_FIELD_NAME(id, field)
+#define DELEGATE_FIELD_NAME_VECTOR(id, field)
+#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, field)
+#define DELEGATE_FIELD_SCOPE_NAME_USE(id, field)
+#define DELEGATE_FIELD_SCOPE_NAME_USE_VECTOR(id, field)
+#define DELEGATE_FIELD_ADDRESS(id, field)
```

Comment:
Interesting idea. Worth thinking about, but I'm not sure as in general you need to be careful about omitting any of them, so it's actually good to be forced to write them out, in a way.

In general the way we've avoided the boilerplate from being an issue is writing higher-level C++ abstractions over this type of code. We could do more of that as an alternative.

Reply:
Yeah, I would definitely want this to be an opt-in mechanism so that by default you are forced to provide all the definitions. For some cases, the code only cares about a single type of thing, and even if we add new types of things in the future, the code still will not care about them.